OSPF: adding dynamically routing capability to VPC

agneya2001 opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)


```
    . Added zone level quagga/ospf config API and UI as addtional tab on zone
    . Added service offering with ospf
    . Added service offering for the tier guest network for VPC
    . Modified system vm template so that quagga is pre-installed
    . Added various configuration files that are required by Quagga
    . Added marvin tests for new APIs and unit test
    . Send quagga config from the server if VPC and tier are dynamically
    routed
    . Configure quagga when a VM of the tier is created/destroyed
    . Added necessary scritps on the vpc router to enable and configure
    quagga and iptable rules
    . Multiple Cidr management and netmask based cidr allocation
    . Added unit tests for cidr management
    . New dialog for create dynamic VPC and netmask as a field
    . Marvin tests to test dynmaic vpc service offering
```

remibergsma commented on setup/db/db/schema-471to480.sql:
```diff
@@ -27,3 +27,10 @@ ALTER TABLE `cloud`.`image_store` MODIFY COLUMN `url` varchar(2048);
 ALTER TABLE `cloud`.`template_store_ref` MODIFY COLUMN `url` varchar(2048);
 ALTER TABLE `cloud`.`volume_store_ref` MODIFY COLUMN `url` varchar(2048);
 ALTER TABLE `cloud`.`volume_store_ref` MODIFY COLUMN `download_url` varchar(2048);
+
```
Please use `schema-481to490.sql` instead as `4.8.0` is already released.

agneya2001 replied:
@remibergsma thank you, will adjust the versions

agneya2001 commented:
Marvin test log for testing ospf: 

nosetests --with-xunit --xunit-file=/tmp/simulator/smoke/test_quota.xml --with-marvin --marvin-config=/root/cloudstack_shapeblue/advanced.cfg /root/cloudstack_shapeblue/test/integration/component/test_ospf.py -s -a tags=advanced,required_hardware=false --zone=Bootcamp --hypervisor=xenserver

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp//MarvinLogs//Mar_17_2016_15_54_43_FRA4VS. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
test_01_dynamic_routing_using_ospf (integration.component.test_ospf.TestCreateDynmaicServiceOffering)
Response quaggaEnabled: true
Response ospfSuperCIDR: 200.100.0.0/16
Response quaggaEnabled: true
Response ospfSuperCIDR: 200.100.0.0/16
Check if the VPC offering is created successfully?
VPC offering is created successfully - Cosmic VPC off-8V5DNP
VPC Offering id = 0976e8a3-87ec-4f47-854d-7fb6433a64a4
Check if the VPC network is created successfully?
VPC network created successfully - Batman VPC-U784HI
VPC batman_vpc id = db59e37c-bd73-4c7b-b9af-6de6f494b4b0
Check if the VPC network is created successfully?
VPC network created successfully - Superman VPC-9OWLC6
VPC superman_vpc id = ffcdfa4d-2af8-43db-b7df-f3abc3e6ce78
Dynamic Network offering id = ffd45aec-9917-471e-8e90-297584963618
Created a network ACL list acl
Added a network ACL 3546e9cc-570f-495a-9bf2-60e1e4b823f2 to ACL list acl
Added a network ACL 02a4dbb2-e29a-4653-9d06-868d0aa6406c to ACL list acl
Verifying list network response to check if network created?
Created batman_network_tier1 with ID: c2eb368f-330c-4e08-b36f-a431e369023f
Created a network ACL list acl
Added a network ACL 48042210-7e6c-4001-97bb-625a36da9584 to ACL list acl
Added a network ACL a96782de-1d30-4650-b896-d27089ef9258 to ACL list acl
Verifying list network response to check if network created?
Created superman_network_tier1 with ID: e037f276-5527-43f5-8916-e97aaf0f4656
Using template 06415103-d802-44de-ba4c-82f1f465dda8 
Using service offering 438477b3-f54d-4d59-a845-520ac91a874f 
Deployed VM 75b13cfa-09d7-4836-a8ba-2b4aa33346f9 in network: c2eb368f-330c-4e08-b36f-a431e369023f
Deployed VM c5bbab5c-108b-4cfe-88d7-f87ae4e27bc3 in network: e037f276-5527-43f5-8916-e97aaf0f4656
Acquired public ip for batman vm 192.168.100.69
Static Nat enabled on ip 192.168.100.69
====Trying SSH Connection: Host:192.168.100.69 User:root                                   Port:22 RetryCnt:60===
===SSH to Host 192.168.100.69 port : 22 SUCCESSFUL===
{Cmd: ping -c 1 200.100.5.101 via Host: 192.168.100.69} {returns: [u'PING 200.100.5.101 (200.100.5.101): 56 data bytes', u'64 bytes from 200.100.5.101: seq=0 ttl=62 time=2.347 ms', u'', u'--- 200.100.5.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 2.347/2.347/2.347 ms']}
[u'PING 200.100.5.101 (200.100.5.101): 56 data bytes', u'64 bytes from 200.100.5.101: seq=0 ttl=62 time=2.347 ms', u'', u'--- 200.100.5.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 2.347/2.347/2.347 ms']
Test successful clearing VMs
Test successful clearing VPC Tiers
Test successful clearing VPCs
=== TestName: test_01_dynamic_routing_using_ospf | Status : SUCCESS ===

===final results are now copied to: /tmp//MarvinLogs/test_ospf_DHP8KX===

agneya2001 commented:
Documentation PR is here https://github.com/apache/cloudstack-docs-admin/pull/36

eriweb commented:
Would you mind running a broader test suite to verify that there are no regressions?

bvbharatk commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 111
 Hypervisor xenserver
 NetworkType Advanced
 Passed=105
 Failed=13
 Skipped=4

**The follwing tests have known issues**
test_vpc_remote_access_vpn
test_vpc_site2site_vpn
test_01_test_vm_volume_snapshot
test_02_vpc_privategw_static_routes
test_03_rvpc_privategw_static_routes
test_04_change_offering_small
ContextSuite context=TestNiciraContoller>:setup
test_01_primary_storage_iscsi
test02_internallb_haproxy_stats_on_all_interfaces
test_01_internallb_roundrobin_1VPC_3VM_HTTP_port80
ContextSuite context=TestDeployVM>:setup
test_04_extract_template
test_04_extract_Iso
test_07_list_default_iso


_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* integration.smoke.test_guest_vlan_range.TestDedicateGuestVlanRange

 * test_dedicateGuestVlanRange Failing since 2 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_deploy_vgpu_enabled_vm
test_06_copy_template
test_06_copy_iso

**Passed test suits:**
integration.smoke.test_deploy_vm_with_userdata.TestDeployVmWithUserData
integration.smoke.test_affinity_groups_projects.TestDeployVmWithAffinityGroup
integration.smoke.test_portable_publicip.TestPortablePublicIPAcquire
integration.smoke.test_over_provisioning.TestUpdateOverProvision
integration.smoke.test_global_settings.TestUpdateConfigWithScope
integration.smoke.test_scale_vm.TestScaleVm
integration.smoke.test_service_offerings.TestCreateServiceOffering
integration.smoke.test_loadbalance.TestLoadBalance
integration.smoke.test_routers.TestRouterServices
integration.smoke.test_reset_vm_on_reboot.TestResetVmOnReboot
integration.smoke.test_snapshots.TestSnapshotRootDisk
integration.smoke.test_deploy_vms_with_varied_deploymentplanners.TestDeployVmWithVariedPlanners
integration.smoke.test_network.TestDeleteAccount
integration.smoke.test_non_contigiousvlan.TestUpdatePhysicalNetwork
integration.smoke.test_deploy_vm_iso.TestDeployVMFromISO
integration.smoke.test_public_ip_range.TestDedicatePublicIPRange
integration.smoke.test_multipleips_per_nic.TestDeployVM
integration.smoke.test_regions.TestRegions
integration.smoke.test_affinity_groups.TestDeployVmWithAffinityGroup
integration.smoke.test_network_acl.TestNetworkACL
integration.smoke.test_pvlan.TestPVLAN
integration.smoke.test_volumes.TestCreateVolume
integration.smoke.test_ssvm.TestSSVMs
integration.smoke.test_nic.TestNic
integration.smoke.test_deploy_vm_root_resize.TestDeployVM
integration.smoke.test_resource_detail.TestResourceDetail
integration.smoke.test_secondary_storage.TestSecStorageServices
integration.smoke.test_vm_life_cycle.TestDeployVM
integration.smoke.test_disk_offerings.TestCreateDiskOffering

agneya2001 commented:
@bvbharatk Thank you for running the marvin tests. Do you think any specific test that is failing on this branch ?

jburwell commented on api/src/com/cloud/network/Network.java:
```diff
@@ -243,7 +245,7 @@ public static Capability getCapability(String capabilityName) {
         Allocated("Indicates the network configuration is in allocated but not setup"), Setup("Indicates the network configuration is setup"), Implementing(
                 "Indicates the network configuration is being implemented"), Implemented("Indicates the network configuration is in use"), Shutdown(
-                "Indicates the network configuration is being destroyed"), Destroy("Indicates that the network is destroyed");
+                        "Indicates the network configuration is being destroyed"), Destroy("Indicates that the network is destroyed");
```
This appears to be an unnecessary formatting change that actually misaligns the code.

jburwell commented on api/src/com/cloud/network/vpc/QuaggaZoneConfig.java:
```diff
@@ -0,0 +1,331 @@
+    public static final String s_quaggaAuthentication = "quaggaauthentication";
+    public static final String s_ospfSuperCIDR = "ospfsupercidr";
+    public static final String s_quaggaPassword = "quaggapassword";
+    public static final String s_quaggaEnabled = "quaggaenabled";
```
Why not collapse these constants into the ``Params`` enumeration with a lookup method to return the enumeration value for the parameter key?

jburwell commented on api/src/com/cloud/network/vpc/QuaggaZoneConfig.java:
```diff
@@ -0,0 +1,331 @@
+public class QuaggaZoneConfig {
+
+    private long zoneId;
+    private String protocol;
```
Assuming that the list of protocols is constrained, why isn't the type an enumeration?

jburwell commented on api/src/com/cloud/network/vpc/QuaggaZoneConfig.java:
```diff
@@ -0,0 +1,331 @@
+    private String helloInterval;
+    private String deadInterval;
+    private String retransmitInterval;
+    private String transitDelay;
```
Assuming ``helloInterval``, ``deadInterval``, ``retransmitInterval`, and ``transitDelay`` are numeric values, why are they represented using ``String``?  Consider refactoring to [Joda-Time](http://www.joda.org/joda-time)'s [Duration](http://www.joda.org/joda-time/apidocs/org/joda/time/Duration.html) to provide a more precise time and conversion semantics.

jburwell commented on api/src/com/cloud/network/vpc/QuaggaZoneConfig.java:
```diff
@@ -0,0 +1,331 @@
+    private String transitDelay;
+    private String authentication;
+    private String password;
+    private String[] ospfSuperCIDRList;
```
Why are we using an ``Array`` instead of a ``List`` or a ``Set``?  Ideally, it seems like we would use a ``Set`` to ensure uniqueness of the CIDRs.

jburwell commented on api/src/com/cloud/network/vpc/QuaggaZoneConfig.java:
```diff
@@ -0,0 +1,331 @@
+    private String authentication;
+    private String password;
+    private String[] ospfSuperCIDRList;
+    private String enabled;
```
Is ``enabled`` representing a ``boolean`` value?  If so, why are using a ``String`` to represent it?

jburwell commented on api/src/com/cloud/network/vpc/VpcProvisioningService.java:
```diff
@@ -26,13 +25,11 @@
-                                         Map serviceCapabilitystList,
-                                         Long serviceOfferingId);
+    public VpcOffering createVpcOffering(String name, String displayText, List<String> supportedServices, Map<String, List<String>> serviceProviders, Map serviceCapabilitystList,
+            Long serviceOfferingId);
```
This appears to be a spurious formatting change that violates our coding standards.

abhinandanprateek replied:
@jburwell 
Have picked up the formatter for eclipse from https://cwiki.apache.org/confluence/display/CLOUDSTACK/Coding+conventions
Have not been fixing the style and formatting by hand but relying on the auto formatting that conforms to apache standard as per the above wiki.

jburwell commented on api/src/com/cloud/network/vpc/VpcProvisioningService.java:
```diff
@@ -26,13 +25,11 @@
-    Pair<List<? extends VpcOffering>,Integer> listVpcOfferings(Long id, String name, String displayText, List<String> supportedServicesStr, Boolean isDefault, String keyword,
-        String state, Long startIndex, Long pageSizeVal);
+    Pair<List<? extends VpcOffering>, Integer> listVpcOfferings(Long id, String name, String displayText, List<String> supportedServicesStr, Boolean isDefault, String keyword,
+            String state, Long startIndex, Long pageSizeVal);
```
This appears to be a spurious formatting change that violates our coding standards.

abhinandanprateek replied:
@jburwell 
Have picked up the formatter for eclipse from https://cwiki.apache.org/confluence/display/CLOUDSTACK/Coding+conventions
Have not been fixing the style and formatting by hand but relying on the auto formatting that conforms to apache standard as per the above wiki.

jburwell commented on api/src/com/cloud/network/vpc/VpcProvisioningService.java:
```diff
@@ -49,4 +46,9 @@ public VpcOffering createVpcOffering(String name, String displayText, List<Strin
      */
     public VpcOffering updateVpcOffering(long vpcOffId, String vpcOfferingName, String displayText, String state);
 
+    Map<String, String> quaggaConfigUpdate(Long id, String protocol, String ospfArea, Short helloInterval, Short deadInterval, Short retransmitInterval, Short transitDelay,
```
Please rename to conform to <verb><noun> method naming idiom (e.g. updateQuaggaConfig).

jburwell commented on api/src/com/cloud/network/vpc/VpcProvisioningService.java:
```diff
@@ -49,4 +46,9 @@ public VpcOffering createVpcOffering(String name, String displayText, List<Strin
+    Map<String, String> quaggaConfigUpdate(Long id, String protocol, String ospfArea, Short helloInterval, Short deadInterval, Short retransmitInterval, Short transitDelay,
+            String authentication, String password, String superCIDR, Boolean enabled);
+
+    Map<String, String> quaggaConfig(Long id);
```
Please rename to conform to <verb><noun> method naming idiom (e.g. updateQuaggaConfig).

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigCmd.java:
```diff
@@ -0,0 +1,73 @@
+import com.cloud.network.vpc.VpcProvisioningService;
+import com.cloud.user.Account;
+
+@APICommand(name = "vpcQuaggaConfig", description = "Captures config informaton for quagga", responseObject = VPCQuaggaConfigResponse.class, since = "4.8.0", requestHasSensitiveInfo = true, responseHasSensitiveInfo = false)
```
The since should be increased to ``4.9.0.`` since ``4.8.0`` has already shipped.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigCmd.java:
```diff
@@ -0,0 +1,73 @@
+
+@APICommand(name = "vpcQuaggaConfig", description = "Captures config informaton for quagga", responseObject = VPCQuaggaConfigResponse.class, since = "4.8.0", requestHasSensitiveInfo = true, responseHasSensitiveInfo = false)
+public class VPCQuaggaConfigCmd extends BaseCmd {
+    public static final Logger s_logger = Logger.getLogger(VPCQuaggaConfigCmd.class.getName());
```
The call to ``getName`` is unnecessary since ``getLogger`` has an override for accepting the ``Class`` instance.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigCmd.java:
```diff
@@ -0,0 +1,73 @@
+@APICommand(name = "vpcQuaggaConfig", description = "Captures config informaton for quagga", responseObject = VPCQuaggaConfigResponse.class, since = "4.8.0", requestHasSensitiveInfo = true, responseHasSensitiveInfo = false)
+public class VPCQuaggaConfigCmd extends BaseCmd {
+    public static final Logger s_logger = Logger.getLogger(VPCQuaggaConfigCmd.class.getName());
+    private static final String s_name = "vpcquaggaconfigresponse";
```
This constant only appears to be used in the ``getName`` method.  Since it used in one place, inline it in ``getName`` and remove the constant.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigUpdateCmd.java:
```diff
@@ -0,0 +1,204 @@
+import com.cloud.network.vpc.VpcProvisioningService;
+import com.cloud.user.Account;
+
+@APICommand(name = "vpcQuaggaConfigUpdate", description = "Captures config informaton for quagga", responseObject = VPCQuaggaConfigResponse.class, since = "4.8.0", requestHasSensitiveInfo = true)
```
The since should be increased to ``4.9.0.`` since ``4.8.0`` has already shipped.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigUpdateCmd.java:
```diff
@@ -0,0 +1,204 @@
+
+    @Override
+    public void execute() {
+        s_logger.info(this.toString());
```
Why are logging the content of this request to ``INFO``?  It seems like it should be ``DEBUG``.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCQuaggaConfigUpdateCmd.java:
```diff
@@ -0,0 +1,204 @@
+@APICommand(name = "vpcQuaggaConfigUpdate", description = "Captures config informaton for quagga", responseObject = VPCQuaggaConfigResponse.class, since = "4.8.0", requestHasSensitiveInfo = true)
+public class VPCQuaggaConfigUpdateCmd extends BaseCmd {
+    public static final Logger s_logger = Logger.getLogger(VPCQuaggaConfigUpdateCmd.class.getName());
+    private static final String s_name = "vpcquaggaconfigresponse";
```
This constant appears to be used only by the ``getName`` method.  Remove the constant and in-line to value.

jburwell commented on api/src/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java:
```diff
@@ -211,7 +201,7 @@ public Boolean getDisplayNetwork() {
     @Override
     public boolean isDisplay() {
-        if(displayNetwork == null)
+        if (displayNetwork == null)
```
Either add the curly braces per coding conventions or reduce a ternary.

jburwell commented on core/src/com/cloud/agent/api/storage/ListTemplateCommand.java:
```diff
@@ -43,6 +49,10 @@ public DataStoreTO getDataStore() {
+    public String getNfsVersion() {
+        return nfsVersion;
+    }
+
```
This looks like a useful change to the storage API, but why is it necessary to implement OSPF?  Seems like it should be in a separate PR.

jburwell commented on core/src/com/cloud/agent/api/storage/PrimaryStorageDownloadCommand.java:
```diff
@@ -87,4 +88,12 @@ public String getPrimaryStorageUrl() {
+
+    public void setNfsVersion(String nfsVersion) {
+        this.nfsVersion = nfsVersion;
+    }
```
This looks like a useful change to the storage API, but why is it necessary to implement OSPF?  Seems like it should be in a separate PR.

jburwell commented on core/src/com/cloud/agent/api/storage/CopyVolumeCommand.java:
```diff
@@ -75,4 +76,12 @@ public boolean toSecondaryStorage() {
+
+    public void setNfsVersion(String nfsVersion) {
+        this.nfsVersion = nfsVersion;
+    }
```
This looks like a useful change to the storage API, but why is it necessary to implement OSPF?  Seems like it should be in a separate PR.

agneya2001 replied:
@jburwell Can you click on the OSPF commit for review, somehow during rebase non ospf commits also got included. I will try to clean them up.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    public enum Protocol {
+        OSPF, BGP
```
Should we include BGP since we don't support it yet?

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+    private Short helloInterval;
+    private Short deadInterval;
+    private Short retransmitInterval;
+    private Short transitDelay;
```
@abhinandanprateek why did you chose to represent these time values as ``Short`` rather than ``Duration``?  My issue with ``Short`` is there is no notion of units (e.g. seconds, milliseconds) etc.  Therefore, the conversion semantics are impossible to determine when just represented as a numeric value.

abhinandanprateek replied:
These values are not consumed by cloudstack and just passed along to quagga config on VR, where these are interpreted by quagga service. Short naturally constraints the value to a reasonable range.

jburwell replied:
@abhinandanprateek the benefit of ``Duration`` is that it carries a unit of measure.  This features ensures that any conversions (e.g. milliseconds to seconds) are reliably performed using well-tested code when rendered to the VR configuration.

abhinandanprateek replied:
@jburwell These values are unused by cloudstack, Short provides a nice range to the values. Since they are passed to VR and used by python scripts, Duration will not give us any added advantage here.

jburwell replied:
@abhinandanprateek but ``short`` lacks any notion of time units.  Is that value in milliseconds, seconds, minutes?  Duration provides this resolution and the methods to convert values reliably between units and getting the value as a ``long`` (e.g. ``long getStandardsSeconds()``, ``long getStandardMinutes()``).

abhinandanprateek replied:
@jburwell lets consider "helloInterval" it is a interval is clear from the name and how it is displayed in UI. Internally it is handled as Short as the MS business logic does not operate on its intrinsic quality of time. Now lets consider that we make it duration as in joda.time pacakge. Now MS business logic need to carry on the overheads of this notion of time by calling the right constructor with additional work required to interpret the parameters and then again doing that conversion when passing this value to virtual router scripts which in turn will dump it in a config file. The tricky part is that this variable conveys notion of time as you rightly put. But the management server business logic need not be aware of this notion of time and can safely transfer the values using Short and thus avoid overheads in conversion back and forth.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+    private Short transitDelay;
+    private Authentication authentication;
+    private String password;
+    private String[] superCIDRList;
```
Given that we are keeping an eye towards IPv6, it seems appropriate to create an immutable class to represent a CIDR with semantics that work for both IPv4 and IPv6 addresses.  This class would deprecate all CIDR related functions in ``NetUtils``.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+        this.authentication = details.get(Params.AUTHENTICATION.name()) == null ? Authentication.MD5
+                : Authentication.valueOf(details.get(Params.AUTHENTICATION.name()));
+        this.password = details.get(Params.PASSWORD.name()) == null ? "" : details.get(Params.PASSWORD.name());
+        String cidr_list = details.get(Params.SUPER_CIDR.name()) == null ? "200.100.0.0/20" : details.get(Params.SUPER_CIDR.name());
```
How was the default value of ``200.100.0.0/2`` determined?  Should this be configuration as a global option?

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+        this.password = details.get(Params.PASSWORD.name()) == null ? "" : details.get(Params.PASSWORD.name());
+        String cidr_list = details.get(Params.SUPER_CIDR.name()) == null ? "200.100.0.0/20" : details.get(Params.SUPER_CIDR.name());
+        this.superCIDRList = cidr_list.split(",");
+        this.enabled = details.get(Params.ENABLED.name()) == null ? Boolean.FALSE : Boolean.valueOf(details.get(Params.ENABLED.name()));
```
1. This ternary operation is repeated throughout this method.  Please extract to a common utility method such as ``getMapValueWithDefault`` and include unit tests.
1. Since the value being retrieved is a ``String``, test the value with ``Strings.isNullOrEmpty``to protect against both null and blank values.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,332 @@
+        this.enabled = details.get(Params.ENABLED.name()) == null ? Boolean.FALSE : Boolean.valueOf(details.get(Params.ENABLED.name()));
+    }
+
+    public void setValues(final String protocol, final String ospfArea, final Short helloInterval, final Short deadInterval, final Short retransmitInterval,
```
Why does this method and ``setDefaultValues`` exist?  It feels like these two methods should be replaced by a builder class that sets default values and provides ``with`` methods for each property's default value to be overridden.  The builder class could be an inner class of ``OSPFZoneConfig``.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigCmd.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public void execute() {
```
Add a check that ``zoneid`` is not null and greater than zero.  Also, add a unit test case to verify the behavior of the ``execute`` method.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigCmd.java:
```diff
@@ -0,0 +1,73 @@
+@APICommand(name = "vpcOSPFConfig", description = "Return zone level ospf configuration", responseObject = VPCOSPFConfigResponse.class, since = "4.9.0", requestHasSensitiveInfo = true, responseHasSensitiveInfo = false)
+public class VPCOSPFConfigCmd extends BaseCmd {
+    public static final Logger s_logger = Logger.getLogger(VPCOSPFConfigCmd.class);
+    private static final String s_name = "vpcospfconfigresponse";
```
This constant is only used in the ``getCommandName`` method.  Why not in-line and simplify the code?

abhinandanprateek replied:
The response name appears in the response, while doing UI or looking at response it is easy to look at the response class and get the top level xml/json level information in that fixed place. Moreover any other response class follows this standard placement, so just followed that convention.

swill commented:
Can you rebase this PR based on the current master so we can resolve the conflicts so we can get CI running against this again.  I want to do a push on the PRs that change the systemvms so we can get a head start on that before the release.

rhtyd commented on tools/appliance/definitions/systemvmtemplate/install_systemvm_packages.sh:
```diff
@@ -71,7 +71,7 @@ function install_packages() {
     libtcnative-1 libssl-dev libapr1-dev \
     python-flask \
-    haproxy \
+    haproxy quagga \
```
@agneya2001 do we only need this additional package? we'll need to build new systemvm template; in that case please change the min. required and we'll need an upgrade path that does that (moves to a 4.9 systemvmtemplate). We'll also need to build/host 4.9 systemvms somewhere.

abhinandanprateek replied:
yes we will need a new system vm template that has quagga preinstalled for ospf functionality to work. Other than that there are no changes and standard template upgrade procedure should apply.

rhtyd replied:
Thanks @abhinandanprateek I'll see what I can do about building the systemvmtemplates

abhinandanprateek commented:
+ TMP=/tmp
+ CLOUDDIR=/root/cloudstack_shapeblue
+ mkdir -p /tmp/simulator/smoke/misc
+ nosetests --with-xunit --xunit-file=/tmp/simulator/smoke/test_quota.xml --with-marvin --marvin-config=/root/cloudstack_shapeblue/advanced.cfg /root/cloudstack_shapeblue/test/integration/component/test_ospf.py -s -a tags=advanced,required_hardware=false --zone=Bootcamp --hypervisor=xenserver

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp//MarvinLogs//Apr_28_2016_10_11_40_GB5OFJ. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
test_01_dynamic_routing_using_ospf (integration.component.test_ospf.TestCreateDynmaicServiceOffering)
Response ospfEnabled: True
Response ospfSuperCIDR: 200.100.0.0/16
Response ospf Enabled: True
Response ospfSuperCIDR: 200.100.0.0/16
Check if the VPC offering is created successfully?
VPC offering is created successfully - Cosmic VPC off-H7FBTX
VPC Offering id = 605c77b9-e377-485d-b2b0-5b67c0f8c20b
Check if the VPC network is created successfully?
VPC network created successfully - Batman VPC-68RBKE
VPC batman_vpc id = bd637cab-976b-49cb-8b2e-24f105b8580e
Check if the VPC network is created successfully?
VPC network created successfully - Superman VPC-C7UIA5
VPC superman_vpc id = cc52d78a-0388-4879-b86a-a34de16107e7
Dynamic Network offering id = a14c863b-c9e0-4ed5-9102-6ceeb370c51b
Created a network ACL list acl
Added a network ACL b9bef6a8-7787-4e57-8352-a6a6f8760c0a to ACL list acl
Added a network ACL dd6e799a-1070-4de8-8670-67adebd3ba89 to ACL list acl
Verifying list network response to check if network created?
Created batman_network_tier1 with ID: 7d0dbf7b-f510-4cc0-ae0e-e0947144a49a
Created a network ACL list acl
Added a network ACL 77d024bd-2970-4123-952f-577483ce6bb5 to ACL list acl
Added a network ACL 5466f0f2-3ba1-43d4-9236-13e547182026 to ACL list acl
Verifying list network response to check if network created?
Created superman_network_tier1 with ID: 1ce4e645-997f-47cf-9903-570ae92f56bb
Using template 7ecce930-30ee-4201-b442-6d376c869b22 
Using service offering 873acd92-328a-4566-afb4-7270650dcfce 
Deployed VM b501f6b0-29f3-4818-80f8-cfbaf78e214f in network: 7d0dbf7b-f510-4cc0-ae0e-e0947144a49a
Deployed VM b675cefe-2014-47bb-a5b3-a67a6507bc12 in network: 1ce4e645-997f-47cf-9903-570ae92f56bb
Acquired public ip for batman vm 192.168.100.69
Static Nat enabled on ip 192.168.100.69
====Trying SSH Connection: Host:192.168.100.69 User:root                                   Port:22 RetryCnt:60===
===SSH to Host 192.168.100.69 port : 22 SUCCESSFUL===
{Cmd: ping -c 1 200.100.5.101 via Host: 192.168.100.69} {returns: [u'PING 200.100.5.101 (200.100.5.101): 56 data bytes', u'64 bytes from 200.100.5.101: seq=0 ttl=62 time=3.703 ms', u'', u'--- 200.100.5.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 3.703/3.703/3.703 ms']}
[u'PING 200.100.5.101 (200.100.5.101): 56 data bytes', u'64 bytes from 200.100.5.101: seq=0 ttl=62 time=3.703 ms', u'', u'--- 200.100.5.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 3.703/3.703/3.703 ms']
{Cmd: ping -c 1 200.100.1.101 via Host: 192.168.100.69} {returns: [u'PING 200.100.1.101 (200.100.1.101): 56 data bytes', u'64 bytes from 200.100.1.101: seq=0 ttl=64 time=0.033 ms', u'', u'--- 200.100.1.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 0.033/0.033/0.033 ms']}
[u'PING 200.100.1.101 (200.100.1.101): 56 data bytes', u'64 bytes from 200.100.1.101: seq=0 ttl=64 time=0.033 ms', u'', u'--- 200.100.1.101 ping statistics ---', u'1 packets transmitted, 1 packets received, 0% packet loss', u'round-trip min/avg/max = 0.033/0.033/0.033 ms']
Test successful clearing VMs
Test successful clearing VPC Tiers
Test successful clearing VPCs
=== TestName: test_01_dynamic_routing_using_ospf | Status : SUCCESS ===

===final results are now copied to: /tmp//MarvinLogs/test_ospf_1POT15===

rhtyd commented:
tag:needlove

rhtyd commented:
SystemVM templates with quagga etc if anyone wants to test --https://packages.shapeblue.com/systemvmtemplate/custom/ospf/

rhtyd commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigCmd.java:
```diff
@@ -0,0 +1,80 @@
+import com.cloud.utils.net.cidr.BadCIDRException;
+
+@APICommand(name = "vpcOSPFConfig", description = "Return zone level ospf configuration", responseObject = VPCOSPFConfigResponse.class, since = "4.9.0", requestHasSensitiveInfo = true, responseHasSensitiveInfo = false)
+public class VPCOSPFConfigCmd extends BaseCmd {
```
Add `since = "4.9.0"`

rhtyd commented on api/src/org/apache/cloudstack/usage/UsageTypes.java:
```diff
@@ -42,6 +42,7 @@
     public static final int VM_DISK_BYTES_READ = 23;
     public static final int VM_DISK_BYTES_WRITE = 24;
     public static final int VM_SNAPSHOT = 25;
+    public static final int DYNAMIC_ROUTING_POLICY = 16;
```
As a new type, please use 26 or above, in case type value = 16 is used by someone else (or externally)

rhtyd commented on client/WEB-INF/classes/resources/messages.properties:
```diff
@@ -2260,3 +2260,29 @@ message.configure.firewall.rules.allow.traffic=Configure the rules to allow Traf
+label.vpc.dynamicrouting.ipv4.netmask=IPv4 Netmask
+
+
+
```
Remove extra newlines (costmetics)

rhtyd commented on client/tomcatconf/commands.properties.in:
```diff
@@ -485,6 +485,10 @@ updateVPCOffering=1
 
+#### VPC OSPF config commands
+vpcOSPFConfig=15
+vpcOSPFConfigUpdate=1
```
To enable the APIs by default, please use `authorized = {RoleType.Admin, ...}` in the @APICommand annotation for these API. Remove changes from this file, we should no longer be changing this file.

abhinandanprateek replied:
@rhtyd is it necessary to do this change for 4.9 or can be done later. I am asking this as I checked some other APIs and did not see the annotation.

rhtyd replied:
@abhinandanprateek the best practice is to add them to API annotation. There are few APIs using these annotations, but I agree that this is not yet popular among API authors.

jburwell replied:
+1 @rhtyd recommendation

abhinandanprateek replied:
In that case the change should include availability version to 4.10 too.

jburwell commented:
@abhinandanprateek please check the Jenkins issues and trigger a rebuild to get this PR green.

abhinandanprateek commented:
@jburwell @swill the PR is in green now on triggering a build.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+    private Short transitDelay;
+    private Authentication authentication;
+    private String password;
+    private CIDR[] superCIDRList;
```
Why is this list of values represented as an array and not a ``List`` or ``Set``?  Is order required?  If not, a ``Set`` would provide uniqueness guarantees.

abhinandanprateek replied:
Uniqness guarantee is not enough by itself, since more complex validations need to run on the CIDR list like the type and overlap , Set would not have provided a major advantage.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+        this.authentication = details.get(Params.AUTHENTICATION.name()) == null ? Authentication.MD5 : Authentication.valueOf(details.get(Params.AUTHENTICATION.name()));
+        this.password = details.get(Params.PASSWORD.name()) == null ? "" : details.get(Params.PASSWORD.name());
+        this.superCIDRList = details.get(Params.SUPER_CIDR.name()) == null ? new CIDR[0] : NetUtils.convertToCIDR(details.get(Params.SUPER_CIDR.name()).split(","));
+        this.enabled = details.get(Params.ENABLED.name()) == null ? Boolean.FALSE : Boolean.valueOf(details.get(Params.ENABLED.name()));
```
There is a repeating group of the following logic:

```details.get(<key>) == null ? <default value> : details.get(<key>)```

Please extract to utility method in CloudStack comment (e.g. ``<K, V> V getOptionalValue(final Map<K, V> map, K key, V default)``) with the appropriate unit tests.  Also, consider using ``Map.containsKey`` to check for key existence.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+        if (superCIDRList != null) {
+            String[] cidr_list = superCIDRList.split(",");
+            for (String cidr : cidr_list) {
+                if (!NetUtils.isValidCIDR(cidr)) {
```
Why not provide a static ``isValid(final String value)`` static method on the ``CIDR`` class?

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+                    throw new InvalidParameterValueException("The super CIDR is not a valid cidr " + cidr);
+                }
+            }
+            CIDR[] v_cidr = NetUtils.convertToCIDR(cidr_list);
```
Why not provide a ``static List<CIDR> fromValues(Iterable<String>)`` factory method on the ``CIDR`` class?

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+                }
+            }
+            CIDR[] v_cidr = NetUtils.convertToCIDR(cidr_list);
+            if (!NetUtils.cidrListConsistency(v_cidr)) {
```
Looking at Line 206 and this line, I am wondering if a CIDR range or collection type would be useful since we operate ordered sets of CIDRs.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+                throw new InvalidParameterValueException("The cidr list is not consistent " + Arrays.toString(cidr_list));
+            }
+            this.superCIDRList = v_cidr;
+        }
```
Lines 261-273 appear to be a repeat of 199-211.  Please to a common method or consider integrating into the ``CIDR`` collection class discussed earlier.

jburwell commented on api/src/com/cloud/network/vpc/OSPFZoneConfig.java:
```diff
@@ -0,0 +1,329 @@
+        }
+    }
+
+    public void setValues(Map<String, String> details) throws BadCIDRException {
```
There appears to be a lot of overlap between this method and ``setDefaultValues``.  What is the difference between these methods?

abhinandanprateek replied:
setDefaultValues bootstraps the config. I have a feeling that to make the API easy on the user, I have created an unnecessary complexity. Like I assume default values when not provided. Instead this method could have been much cleaner without any assumptions of default and ensuring that all the values are provided all the time.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_helloInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for HelloInterval timer value.(default 10)")
+    private Short helloInterval;
+
+    @Parameter(name = OSPFZoneConfig.s_deadInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for RouterDeadInterval timer value used for Wait Timer and Inactivity Timer.(default 40)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_area, type = CommandType.STRING, required = false, description = "Specify the OSPF Area ID")
+    private String area;
+
+    @Parameter(name = OSPFZoneConfig.s_helloInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for HelloInterval timer value.(default 10)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_deadInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for RouterDeadInterval timer value used for Wait Timer and Inactivity Timer.(default 40)")
+    private Short deadInterval;
+
+    @Parameter(name = OSPFZoneConfig.s_retransmitInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for RxmtInterval timer value.(default 5)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

rhtyd replied:
I'm not sure how a default value is applicable here; but in general if an API author wants to enforce a default value, it's doable either from the getter() or set the default value to the field like `private Short retransmitInterval = 10;`. If the API user sends us the value, that value will overwrite the default value of this field, alternatively we can do a conditional check in the getter().

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_retransmitInterval, type = CommandType.SHORT, required = false, description = "Set number of seconds for RxmtInterval timer value.(default 5)")
+    private Short retransmitInterval;
+
+    @Parameter(name = OSPFZoneConfig.s_transitDelay, type = CommandType.SHORT, required = false, description = "Set number of seconds for InfTransDelay value.(default 1)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_transitDelay, type = CommandType.SHORT, required = false, description = "Set number of seconds for InfTransDelay value.(default 1)")
+    private Short transitDelay;
+
+    @Parameter(name = OSPFZoneConfig.s_authentication, type = CommandType.STRING, required = false, description = "Md5 or PlainText (default MD5)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = ApiConstants.ZONE_ID, type = CommandType.UUID, entityType = ZoneResponse.class, required = true, description = "the ID of the Zone")
+    private Long zoneid;
+
+    @Parameter(name = OSPFZoneConfig.s_protocol, type = CommandType.STRING, required = false, description = "the password used to secure inter ospf communication(default ospf)")
```
Specify the default value in the ``default`` parameter of the ``Parameter`` annotation

abhinandanprateek replied:
There is no default value in Parameter annotation. How will that work, like if in request a field is empty fill it with default value if given ? I think it may not be very useful.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_protocol, type = CommandType.STRING, required = false, description = "the password used to secure inter ospf communication(default ospf)")
+    private String protocol;
+
+    @Parameter(name = OSPFZoneConfig.s_area, type = CommandType.STRING, required = false, description = "Specify the OSPF Area ID")
```
Is the OSPF area ID required for proper configuration?

abhinandanprateek replied:
yes, in case the ospf is enabled.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_superCIDR, type = CommandType.STRING, required = false, description = "The super zone level CIDR for ospf enabled VPCs")
+    private String superCIDR;
+
+    @Parameter(name = OSPFZoneConfig.s_enabled, type = CommandType.BOOLEAN, required = false, description = "The flag to enable or disable ospf for this zone")
```
Is there a reasonable default value for this parameter?

abhinandanprateek replied:
The reasonable default values are provided in setDefault method that is used to bootstrap the values for first time.

jburwell commented on api/src/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigUpdateCmd.java:
```diff
@@ -0,0 +1,207 @@
+    @Parameter(name = OSPFZoneConfig.s_password, type = CommandType.STRING, required = false, description = "The password used to secure inter quagga communication")
+    private String password;
+
+    @Parameter(name = OSPFZoneConfig.s_superCIDR, type = CommandType.STRING, required = false, description = "The super zone level CIDR for ospf enabled VPCs")
```
Is a super CIDR required for proper OSPF configuration?

jburwell commented on api/src/org/apache/cloudstack/api/command/user/vpc/CreateVPCCmd.java:
```diff
@@ -143,7 +149,17 @@ public Boolean getDisplayVpc() {
+            }
+        } else {
+            vpc = _vpcService.createVpc(getZoneId(), getVpcOffering(), getEntityOwnerId(), getVpcName(), getDisplayText(), getCidr(), getNetworkDomain(), getDisplayVpc());
+        }
```
Why not perform this check in the ``vpcService.createVpc`` method?  This construct feels like a leak of the business rule.  It might make sense to create a VPC interface and factory that creates the appropriate type (static vs dynamic VPC).  The ``_vpcService`` could then provide polymorphic implementations of ``createVPC`` -- letting the compiler handle the variance for service clients.

jburwell commented on api/test/org/apache/cloudstack/api/command/admin/vpc/VPCOSPFConfigCmdTest.java:
```diff
@@ -0,0 +1,72 @@
+        MockitoAnnotations.initMocks(this);
+        Field f = VPCOSPFConfigUpdateCmd.class.getDeclaredField("_vpcProvSvc");
+        f.setAccessible(true);
+        f.set(vpcOSPFConfig, vpcProvSvc);
```
Can these mocks be setup through the ``SpringJUnitRunner``?  While effective, the reflection approach is non-intuitive.  and potentially brittle (e.g. someone refactors the attribute name).

jburwell commented on core/src/com/cloud/agent/api/routing/QuaggaConfigCommand.java:
```diff
@@ -0,0 +1,86 @@
+    private CIDR[] tierCidrs;
+    private OSPFZoneConfig zoneConfig;
+
+    public QuaggaConfigCommand(final String routerPublicIp, final String routerName, final String vpnCidr, final CIDR[] tierCidrs, final OSPFZoneConfig qzc) {
```
Further elaborating on my previous comments about  a CIDR collection type, would it make sense to have a ``Tiers`` class that implements ``Iterable`` and composes ``CIDR``?

abhinandanprateek replied:
The tier cidrs are provided by the user. I think it would have mode more sense if they were being generated automatically from a given cidr. So a method that breaks a given cidr using a netmask into tiers would be a good application of what you are suggesting.

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+        final QuaggaRule quaggaRule = new QuaggaRule(zebra.toString(), ospfd.toString(), tmpCfgFilePath, tmpCfgFileName, vrIp);
+
+        return generateConfigItems(quaggaRule);
+    }
```
This method appears to work with the following concerns:

1. A common zerba/ospfd configuration format (lines of space separated values ending with a ``,`` with child items idented with a two (2) spaces).  It also includes conversion of various types to ``String``.
1. Generation of a zebra configuration
1. Generation of an ospfd configuration

The mix of these concerns has created repeating groups and a method that is larger than necessary.  Please consider decomposing as follows:

1. Create a ``QuaggaConfFileBuilder`` class that encapsulates the configuration file format concern.  This class could methods such as ``QuaggaConfFileBuilder withDirective(String... elements)`` that joins the passed ``elements`` with a space and appends a comma; ``QuaggaConfFileBuilder addChildDirective()`` which returns a nested builder to represent the indentation hierarchy; etc.
1. Split the creation of the zebra and ospf configurations into separate private methods or individual small classes.  The decision as to which approach would be most effective would depend on how many rules/data type conversions need to be represented.

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+        ospfd.append("log file /var/log/quagga/ospfd.log").append(",");
+
+        final String tmpCfgFilePath = "/etc/quagga/";
+        final String tmpCfgFileName = "ospfd.conf.new." + String.valueOf(System.currentTimeMillis());
```
Why isn't the zebra configuration backed up?

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+        zebra.append("hostname " + vrName).append(",");
+        zebra.append("interface eth1").append(",");
+        zebra.append(" description link to area 0").append(",");
+        zebra.append("  ip address " + vrIp + "/24").append(",");
```
Is this representation IPv4/IPv6 dual stack compatible?

abhinandanprateek replied:
No, it is ipv4 specific.

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+        ospfd.append("  ospf router-id " + vrIp).append(",");
+        ospfd.append("  redistribute connected").append(",");
+        ospfd.append("  no passive-interface eth1").append(",");
+        ospfd.append("  network " + vrIp + "/24 area ").append(qzc.getOspfArea()).append(",");
```
Is this representation IPv4/IPv6 dual stack compatible?

abhinandanprateek replied:
No, it is ipv4 specific.

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+        ospfd.append("router ospf").append(",");
+        ospfd.append("  ospf router-id " + vrIp).append(",");
+        ospfd.append("  redistribute connected").append(",");
+        ospfd.append("  no passive-interface eth1").append(",");
```
Is it safe to assume the ``eth1`` device if other services are being configured on the VR?

abhinandanprateek replied:
First 2 nics on VR are for specific purpose to CS. If there are more network service there will be more nics. If there are multiple public networks there will be more public nics.

jburwell commented on core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java:
```diff
@@ -0,0 +1,93 @@
+    public List<ConfigItem> generateConfig(final NetworkElementCommand cmd) {
+        final QuaggaConfigCommand command = (QuaggaConfigCommand)cmd;
+
+        final String vrIp = command.getRouterPublicIp();
```
Why isn't ``vrIp`` being represented as ``IpAddress`` for dual stack encapsulation?

abhinandanprateek replied:
Can be done, but it will be a bigger restructuring exercise as router ip touches almost all networking code.

jburwell commented on engine/components-api/src/com/cloud/network/vpc/VpcManager.java:
```diff
@@ -161,8 +161,9 @@
 
     List<PrivateGateway> getVpcPrivateGateways(long vpcId);
+
+    CIDR[] getZoneSuperCidrList(long zoneId) throws BadCIDRException;
```
Why is the list of super CIDRs being represented as an array instead of a ``List``, ``Set``, or custom collected type for ``CIDR``s?

jburwell commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -1445,28 +1466,31 @@ public void prepareAllNicsForMigration(final VirtualMachineProfile vm, final Dep
-            if(guru instanceof NetworkMigrationResponder){
-                if(!((NetworkMigrationResponder) guru).prepareMigration(profile, network, vm, dest, context)){
-                    s_logger.error("NetworkGuru "+guru+" prepareForMigration failed."); // XXX: Transaction error
+            if (guru instanceof NetworkMigrationResponder) {
```
What if ``guru`` is ``null``?  Should we have a sanity ``Preconditions.checkState`` to avoid an NPE?

abhinandanprateek replied:
I think it is some formatting change as there are no ospf related changes in this part of the code.
Guru is a implementation for network. Though there is always a possibility for it to go null, it will not happen as cloudstack ships with default implementations.

jburwell commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -1445,28 +1466,31 @@ public void prepareAllNicsForMigration(final VirtualMachineProfile vm, final Dep
-                    s_logger.error("NetworkGuru "+guru+" prepareForMigration failed."); // XXX: Transaction error
+            if (guru instanceof NetworkMigrationResponder) {
+                if (!((NetworkMigrationResponder)guru).prepareMigration(profile, network, vm, dest, context)) {
+                    s_logger.error("NetworkGuru " + guru + " prepareForMigration failed."); // XXX: Transaction error
```
Should this method continue if the ``prepareMigration`` fails?  If not, should it simply return or throw an exception?

jburwell commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -1445,28 +1466,31 @@ public void prepareAllNicsForMigration(final VirtualMachineProfile vm, final Dep
-                            s_logger.error("NetworkElement "+element+" prepareForMigration failed."); // XXX: Transaction error
+                    if (element instanceof NetworkMigrationResponder) {
+                        if (!((NetworkMigrationResponder)element).prepareMigration(profile, network, vm, dest, context)) {
+                            s_logger.error("NetworkElement " + element + " prepareForMigration failed."); // XXX: Transaction error
```
Should this method continue if the ``prepareMigration`` call fails?  If not, should it return or throw an exception?

jburwell commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -1476,22 +1500,22 @@ public void prepareAllNicsForMigration(final VirtualMachineProfile vm, final Dep
                     //Nic details are not available in DB
                     //Create nic profile for migration
-                    s_logger.debug("Creating nic profile for migration. BroadcastUri: "+broadcastUri.toString()+" NetworkId: "+ntwkId+" Vm: "+vm.getId());
+                    s_logger.debug("Creating nic profile for migration. BroadcastUri: " + broadcastUri.toString() + " NetworkId: " + ntwkId + " Vm: " + vm.getId());
```
Wrap this statement in a ``s_logger.isDebugEnabled`` call to avoid unnecessary string creation when ``DEBUG`` logging is not enabled.

jburwell commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:
```diff
@@ -2024,8 +2047,25 @@ public Network createGuestNetwork(final long networkOfferingId, final String nam
+            else {
+                if (!NetUtils.validateGuestCidr(cidr)) {
+                    throw new InvalidParameterValueException("Virtual Guest Cidr " + cidr + " is not RFC1918 compliant");
+                }
```
Why not wrap this validation logic in a static method on the ``CIDR`` interface?

jburwell commented on engine/schema/src/com/cloud/dc/DataCenterVO.java:
```diff
@@ -365,6 +366,18 @@ public void setLocalStorageEnabled(boolean enabled) {
 
     @Override
+    public boolean isDynamicRoutingEnabled() {
+        String dynr = getDetail(OSPFZoneConfig.Params.ENABLED.name());
```
There is a common pattern across this patch where ``getDetail`` is called and then the returned value is checked.  If the return value is ``null``, a default value is returned.  Please consider adding a ``getOptionalDetail(key, defaultValue)`` method that encapsulates this logic.  If this method is added, please add a unit test to verify its behavior.

jburwell commented on engine/schema/src/com/cloud/network/dao/PhysicalNetworkServiceProviderVO.java:
```diff
@@ -308,15 +331,21 @@ public void setEnabledServices(List<Service> services) {
+                + sourcenatServiceProvided + ", lbServiceProvided=" + lbServiceProvided + ", staticnatServiceProvided=" + staticnatServiceProvided
+                + ", portForwardingServiceProvided=" + portForwardingServiceProvided + ", userdataServiceProvided=" + userdataServiceProvided + ", securitygroupServiceProvided="
+                + securitygroupServiceProvided + ", networkAclServiceProvided=" + networkAclServiceProvided + ", dynamicRoutingProvided=" + dynamicRoutingProvided + ", removed="
+                + removed + "]";
```
Given number of concatenations, please refactor to use a ``StringBuilder`` in order to reduce the number of strings created.

jburwell commented on engine/schema/src/com/cloud/network/vpc/dao/VpcServiceMapDaoImpl.java:
```diff
@@ -62,7 +62,31 @@ protected VpcServiceMapDaoImpl() {
+        List<VpcServiceMapVO> vpcServices = listBy(sc);
+
+        if (services != null) {
+            if (vpcServices.size() == services.length) {
```
What if the size of ``vpcServices`` and ``services`` are equal, but they contain different elements?

abhinandanprateek replied:
listBy is going to query the services that are send to the function. So if all the services queried for are present the dao should return that many rows.

jburwell commented on engine/schema/src/com/cloud/network/vpc/dao/VpcServiceMapDaoImpl.java:
```diff
@@ -62,7 +62,31 @@ protected VpcServiceMapDaoImpl() {
+            if (vpcServices.size() == services.length) {
+                return true;
+            }
+        } else if (!vpcServices.isEmpty()) {
```
Since this condition and the previous both return ``true``, why not combine them?

jburwell commented on plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java:
```diff
@@ -138,7 +138,8 @@ public QuotaTariffResponse createQuotaTariffResponse(QuotaTariffVO tariff) {
                 AccountVO account = _accountDao.findById(quotaAccount.getId());
-                if (account == null) continue;
+                if (account == null)
+                    continue;
```
Please wrap in curly braces to conform with our coding standards.

jburwell commented on server/src/com/cloud/network/NetworkServiceImpl.java:
```diff
@@ -2111,8 +2098,9 @@ public Network updateGuestNetwork(final long networkId, String name, String disp
-                _entityMgr.findById(NetworkOffering.class, networkOfferingId), network.getPhysicalNetworkId()) : new HashMap<String, String>();
+        final Map<String, String> newSvcProviders = networkOfferingChanged
+                ? _networkMgr.finalizeServicesAndProvidersForNetwork(_entityMgr.findById(NetworkOffering.class, networkOfferingId), network.getPhysicalNetworkId())
+                : new HashMap<String, String>();
```
Recognizing it is out of scope of this patch, but would it be possible to change ``new HashMap<String, String>()`` to ``Collections.emptyMap()`` to avoid unnecessary object creation when ``networkOfferingChanged`` is ``false``.

jburwell commented on server/src/com/cloud/network/element/VpcVirtualRouterElement.java:
```diff
@@ -482,8 +474,8 @@ public boolean applyIps(final Network network, final List<? extends PublicIpAddr
-                s_logger.debug(getName() + " element doesn't need to associate ip addresses on the backend; VPC virtual " + "router doesn't exist in the network "
-                        + network.getId());
+                s_logger.debug(
+                        getName() + " element doesn't need to associate ip addresses on the backend; VPC virtual " + "router doesn't exist in the network " + network.getId());
```
While this change is the result of a reformat, would it be possible to wrap this logging statement in a ``s_logger.isDebugEnabled()`` check to avoid unnecessary string concatenation when ``DEBUG`` logging is not enabled?

jburwell commented on server/src/com/cloud/network/router/CommandSetupHelper.java:
```diff
@@ -1048,4 +1052,33 @@ protected String getGuestDhcpRange(final NicProfile guestNic, final Network gues
+        final DataCenterVO dcVo = _dcDao.findById(router.getDataCenterId());
+        cmd.setAccessDetail(NetworkElementCommand.ZONE_NETWORK_TYPE, dcVo.getNetworkType().toString());
+
+        cmds.addCommand(cmd);
```
Create a copy of the ``cmds`` list, add to it, and return it to avoid creating unnecessary side effects for the caller.

abhinandanprateek replied:
Caller is depending on the fact that the method adds commands to the ref he has passed.

jburwell commented on server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:
```diff
@@ -198,14 +199,26 @@ public boolean removeVpcRouterFromGuestNetwork(final VirtualRouter router, final
+                }
+                catch (BadCIDRException ex){
+                    s_logger.warn("Unable to configure the VPC network for dynamic routing " + ex);
+                    throw new ResourceUnavailableException("Unable to configure the VPC network for dynamic routing ," + " virtual router " + router + " is not in the right state", DataCenter.class,
```
Why are two static string values being concatenated?

jburwell commented on server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:
```diff
@@ -455,16 +471,26 @@ protected void finalizeNetworkRulesForNetwork(final Commands cmds, final DomainR
                     if (networkACLs != null && !networkACLs.isEmpty()) {
-                        s_logger.debug("Found " + networkACLs.size() + " network ACLs to apply as a part of VPC VR " + domainRouterVO + " start for guest network id=" + guestNetworkId);
+                        s_logger.debug(
+                                "Found " + networkACLs.size() + " network ACLs to apply as a part of VPC VR " + domainRouterVO + " start for guest network id=" + guestNetworkId);
```
While I recognize this change is due a formatting change, would it be possible to wrap this log statement in an ``s_logger.isDebugEnabled()`` check to avoid unnecessary string concatenation when ``DEBUG`` logging is not enabled?

jburwell commented on server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:
```diff
@@ -455,16 +471,26 @@ protected void finalizeNetworkRulesForNetwork(final Commands cmds, final DomainR
+                    try {
+                        _commandSetupHelper.createQuaggaConfigCommand(domainRouterVO, vpc.getId(), cmds);
+                    } catch (BadCIDRException ex) {
+                        s_logger.debug(ex);
```
Why log to ``DEBUG`` instead of ``ERROR``?

jburwell commented on server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:
```diff
@@ -455,16 +471,26 @@ protected void finalizeNetworkRulesForNetwork(final Commands cmds, final DomainR
+                        _commandSetupHelper.createQuaggaConfigCommand(domainRouterVO, vpc.getId(), cmds);
+                    } catch (BadCIDRException ex) {
+                        s_logger.debug(ex);
+                        throw new CloudRuntimeException("The cidr for dynamic routing is bad " + ex);
```
Please add the ``ex`` as the cause to provide a complete stack trace for debugging.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -309,6 +314,19 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
                 }
+
+                if (_vpcOffDao.findByUniqueName(VpcOffering.routedVPCOfferingName) == null) {
+                    s_logger.debug("Creating default routed VPC offering " + VpcOffering.routedVPCOfferingName);
```
Please wrap this log statement in a ``isDebugEnabled()`` check to avoid unnecessary string concatenation when ``DEBUG`` logging is not enabled.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -309,6 +314,19 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
+
+                    final Map<Service, Set<Provider>> svcProviderMap = new HashMap<Service, Set<Provider>>();
+                    final Set<Provider> defaultProviders = new HashSet<Provider>();
+                    defaultProviders.add(Provider.VPCVirtualRouter);
```
Consider replacing lines 321-323 with Guava's ``SetMultiMap``

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -501,7 +517,7 @@ private void validateConnectivtyServiceCapabilities(final Set<Provider> provider
 
             while (iter.hasNext()) {
-                final HashMap<String, String> svcCapabilityMap = (HashMap<String, String>) iter.next();
+                final HashMap<String, String> svcCapabilityMap = (HashMap<String, String>)iter.next();
```
Please type to the ``Map`` interface rather than ``HashMap`` to reduce coupling.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -529,14 +545,14 @@ private boolean findCapabilityForService(final Map serviceCapabilitystList, fina
             final Iterator iter = serviceCapabilitystList.values().iterator();
             while (iter.hasNext()) {
-                final HashMap<String, String> currentCapabilityMap = (HashMap<String, String>) iter.next();
+                final HashMap<String, String> currentCapabilityMap = (HashMap<String, String>)iter.next();
```
Please type to the ``Map`` interface rather than ``HashMap`` to reduce coupling.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -705,6 +744,30 @@ public boolean deleteVpcOffering(final long offId) {
+        Map<String, String> details = dc.getDetails();
+        qzc.setDefaultValues(details);
+        qzc.setValues(protocol, ospfArea, helloInterval, deadInterval, retransmitInterval, transitDelay, authentication, quaggaPassword, superCIDR, enabled);
+        details = qzc.getValues();
```
Given the number of parameters on ``OSPFZoneConfig.setValues`` with the sequence of calls on lines 754-756, please consider extracting the creation of ``OSPFZoneConfig`` instances to a builder class as a static inner class on ``OSPFZoneConfig``.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -705,6 +744,30 @@ public boolean deleteVpcOffering(final long offId) {
+        details = qzc.getValues();
+        dc.setDetails(details);
+        _dcDao.saveDetails(dc);
+        s_logger.info("Quagga config saved to datacenter " + zoneId);
```
Please include the contents of the config in this log message.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+            CIDR[] cidr_list = NetUtils.convertToCIDR(str_cidr_list);
+            if (!NetUtils.cidrListConsistency(cidr_list)) {
+                throw new InvalidParameterValueException("The cidr list is not consistent " + Arrays.toString(cidr_list));
+            }
```
As noted in earlier comments, please consider a CIDR custom collection type and encapsulation of this logic in it.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+        // get all the used cidrs
+        List<CIDR> usedSubnets = new ArrayList<CIDR>();
+        for (Vpc vpc : _vpcDao.listAll()) {
+            if (_vpcSrvcDao.areServicesSupportedInVpc(vpc.getId(), Service.VPCDynamicRouting)) {
```
For performance reasons, calling a query in a loop should be avoided.  Is it possible to refactor this ``for`` loop into a join?

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+    public synchronized Vpc createDynamicVpc(final long zoneId, final long vpcOffId, final long vpcOwnerId, final String vpcName, final String displayText, final String netmask,
+            String networkDomain, final Boolean displayVpc) throws ResourceAllocationException, BadCIDRException {
+        if (netmask != null) {
+            if (!NetUtils.isValidNetmask(netmask)) {
```
Please consider implementing this validation on the as a static method on the ``Netmask`` interface for IPv4/IPv6 dual stack support.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+            final CIDR[] superCidr = getZoneSuperCidrList(zoneId);
+            final List<CIDR> usedSubnets = getAllVpcCidrs();
+            //find cidr that does not overlap with the used cidrs
+            List<CIDR> allSubnets = NetUtils.getAllSubnets(superCidr, netmask);
```
Consider moving this method to the ``CIDR`` class.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+            //check for overlap with usedSubnets
+            CIDR unused_cidr = null;
+            for (CIDR subnet : allSubnets) {
+                if (NetUtils.isCidrOverlap(subnet, usedSubnets)) {
```
Consider implementing this method on the ``CIDR`` class.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -747,6 +810,67 @@ public VpcOffering updateVpcOffering(final long vpcOffId, final String vpcOfferi
+                    unused_cidr = subnet;
+                    break;
+                }
+            }
```
Consider moving this for loop to ``CIDR.findUnusedSubnet`` method and adding a unit test for it.

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -809,15 +945,22 @@ public Vpc createVpc(final long zoneId, final long vpcOffId, final long vpcOwner
+            final CIDR[] superCidr = getZoneSuperCidrList(vpc.zoneId);
+
+            if (_vpcOffServiceDao.areServicesSupportedByNetworkOffering(vpc.getVpcOfferingId(), Service.VPCDynamicRouting)) {
+                if (!NetUtils.validateGuestCidrForOSPF(cidr, superCidr)) {
```
Can this validation be moved to the ``CIDR`` interface to support IPv4/IPv6 dual stack operation?

jburwell commented on server/src/com/cloud/network/vpc/VpcManagerImpl.java:
```diff
@@ -809,15 +945,22 @@ public Vpc createVpc(final long zoneId, final long vpcOffId, final long vpcOwner
+                if (!NetUtils.validateGuestCidrForOSPF(cidr, superCidr)) {
+                    throw new InvalidParameterValueException("Guest Cidr " + cidr + " is not within Zone superCIDR " + Arrays.toString(superCidr));
+                }
+            } else if (!NetUtils.validateGuestCidr(cidr)) {
```
Can this validation be moved to the ``CIDR`` interface to support IPv4/IPv6 dual stack operation?

jburwell commented on server/src/com/cloud/server/ConfigurationServerImpl.java:
```diff
@@ -1236,18 +1218,34 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
+                for (Map.Entry<Service, Provider> entry : defaultVpcNetworkOfferingProviders.entrySet()) {
+                    NetworkOfferingServiceMapVO offService = new NetworkOfferingServiceMapVO(defaultNetworkOfferingForVpcNetworks.getId(), entry.getKey(), entry.getValue());
+                    _ntwkOfferingServiceMapDao.persist(offService);
+                    s_logger.trace("Added service for the network offering: " + offService);
```
Please wrap this log statement in a ``isTraceEnabled()`` check to avoid unnecessary string concatenation when ``TRACE`` logging is not enabled.

jburwell commented on server/src/org/apache/cloudstack/network/topology/AdvancedNetworkVisitor.java:
```diff
@@ -212,4 +214,20 @@ public boolean visit(final AdvancedVpnRules vpnRules) throws ResourceUnavailable
+        try {
+            _commandSetupHelper.createQuaggaConfigCommand(router, quaggaRules.getVpcId(), cmds);
+        } catch (BadCIDRException ex) {
+            s_logger.warn("Failed to create Quagga config due to faulty cidr ");
```
Please log the exception to support operational debugging.

jburwell commented on server/src/org/apache/cloudstack/network/topology/BasicNetworkVisitor.java:
```diff
@@ -316,4 +317,9 @@ public boolean visit(final StaticRoutesRules staticRoutesRules) throws ResourceU
+
+    @Override
+    public boolean visit(final QuaggaRules vpnRules) throws ResourceUnavailableException {
+        throw new CloudRuntimeException("AdvancedVpnRules not implemented in Basic Network Topology.");
```
Consider using the ``OperationNotSupported`` exception provided by the Java runtime for this purpose.

abhinandanprateek replied:
CloudRuntimeException is UI friendly and is thrown to pop up the error message on CS UI. It is also consistent with other similar methods. I think if we find strong reasons to change it we should do it for all the other similar methods too.

jburwell commented on systemvm/patches/debian/config/opt/cloud/bin/cs/CsQuagga.py:
```diff
@@ -0,0 +1,85 @@
+        ospfd_conf = CsFile(OSPFD_CONF)
+        if not ospfd_conf.compare(ospfd_conf_new):
+            CsHelper.copy(OSPFD_CONF_NEW, OSPFD_CONF)
+            changed = True
```
Lines 43-52 and 55-64 appear to perform identical logic.  Consider extracting a generic, private function such as ``create_conf_file(filename, new_config)`` that returns (boolean, CsFile).  Lines 36-40 could also be collapsed into this function.

jburwell commented on systemvm/patches/debian/config/opt/cloud/bin/cs/CsQuagga.py:
```diff
@@ -0,0 +1,85 @@
+                CsHelper.service("quagga", "restart")
+            else:
+                logging.debug("CsQuagga:: will reload Quagga!")
+                CsHelper.service("quagga", "reload")
```
Where is zebra restarted?

abhinandanprateek replied:
In debian quagga service will restart both ospfd and zebra.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+          # 2) Create ACL
+        self.batman_aclgroup = NetworkACLList.create(apiclient=self.apiclient, services={}, name="acl", description="acl", vpcid=self.batman_vpc.id)
+        self.assertIsNotNone(self.batman_aclgroup, "Failed to create NetworkACL list")
+        self.logger.debug("Created a network ACL list %s" % self.batman_aclgroup.name)
```
Consider extracting lines 311-313 to private function to reduce the complexity of this test method.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+                                                   traffictype='Ingress' )
+        
+        self.assertIsNotNone(self.batman_aclitem2, "Network failed to aclItem")
+        self.logger.debug("Added a network ACL %s to ACL list %s" % (self.batman_aclitem2.id, self.batman_aclgroup.name))
```
Consider extracting lines 315-325 to a private method to reduce the complexity of this test method.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+            vpcid=self.batman_vpc.id
+        )
+        self.validate_network(self.batman_network_tier1, self.batman_vpc.id)
+        self.logger.debug("Created batman_network_tier1 with ID: %s" % self.batman_network_tier1.id)
```
Consider extracting lines 327-345 to a private method to reduce the complexity of this test method.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+        # 2) Create ACL
+        self.superman_aclgroup = NetworkACLList.create(apiclient=self.apiclient, services={}, name="acl", description="acl", vpcid=self.superman_vpc.id)
+        self.assertIsNotNone(self.superman_aclgroup, "Failed to create NetworkACL list")
+        self.logger.debug("Created a network ACL list %s" % self.superman_aclgroup.name)
```
Lines 349-352 appear to be identical to 311-313 -- the previously suggested extraction could be used here as well.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+        self.superman_aclitem2 =  NetworkACL.create( self.apiclient, aclid=self.superman_aclgroup.id, services=self.services["natrule"],
+                                        traffictype='Ingress' )
+        self.assertIsNotNone(self.superman_aclitem2, "Network failed to aclItem")
+        self.logger.debug("Added a network ACL %s to ACL list %s" % (self.superman_aclitem2.id, self.superman_aclgroup.name))
```
Lines 354-364 appear similar to lines 315-325 -- the previously suggested extraction could be used here as well.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+            vpcid=self.superman_vpc.id
+        )
+        self.validate_network(self.superman_network_tier1, self.superman_vpc.id)
+        self.logger.debug("Created superman_network_tier1 with ID: %s" % self.superman_network_tier1.id)
```
Lines 366-385 appear similar to lines 327-345 -- the previously suggested extraction could be used here as well.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+            ipaddress="200.100.5.101",
+            networkids=self.superman_network_tier1.id
+        )
+        self.logger.debug("Deployed VM %s in network: %s" % (self.superman_vm_1.id, self.superman_network_tier1.id))
```
Lines 404-413 and lines 415-424 appear similar.  Consider extracting to a private function to reduce the complexity of this test method.

jburwell commented on test/integration/component/test_ospf.py:
```diff
@@ -0,0 +1,477 @@
+        self.logger.debug("Test successful clearing VPCs")
+        self.cleanup.append(self.superman_vpc)
+        self.cleanup.append(self.batman_vpc)
+        self.cleanup.append(self.vpc_off)
```
Lines 462-473 only appear to add resources to a list.  How does this affect cleanup of the resources?

abhinandanprateek replied:
the resources added to the cleanup list are picked up by Marvin when tearDown is called.

jburwell commented on test/integration/smoke/test_ospf_zone_config.py:
```diff
@@ -0,0 +1,165 @@
+        return
+
+    # Check vpcQuaggaConfigCmd positive test case
+    #VPCQuaggaConfigUpdateCmd [zoneid=1, quaggaProtocol=null, ospfArea=null, quaggaHelloInterval=null, quaggaDeadInterval=null, quaggaRetransmitInterval=null, quaggaTransitDelay=null, quaggaAuthentication=null, quaggaPassword=junk, ospfSuperCIDR=null, quaggaEnabled=null]
```
Please remove commented code to avoid accumulation of cruft.

jburwell commented on test/integration/smoke/test_ospf_zone_config.py:
```diff
@@ -0,0 +1,165 @@
+
+        
+    # Check vpcQuaggaConfigCmd positive test case
+    #VPCQuaggaConfigUpdateCmd [zoneid=1, quaggaProtocol=null, ospfArea=null, quaggaHelloInterval=null, quaggaDeadInterval=null, quaggaRetransmitInterval=null, quaggaTransitDelay=null, quaggaAuthentication=null, quaggaPassword=junk, ospfSuperCIDR=null, quaggaEnabled=null]
```
Please remove commented code to avoid accumulation of cruft.

jburwell commented on test/integration/smoke/test_ospf_zone_config.py:
```diff
@@ -0,0 +1,165 @@
+        
+        
+        self.assertEqual(
+                         response.quaggaPassword, "6P33lsadEbmiJ7ZN7gK8vylOY2dTza3rAuY8UFboH7I=:Xb1G4VhzBkSXmTa8vDqExQQo8PiIUUSNtpxmdOkX1z8="
```
Consider extracting the password to a constant ensure consistency and increase clarity

jburwell commented on test/integration/smoke/test_ospf_zone_config.py:
```diff
@@ -0,0 +1,165 @@
+                         )
+        
+        self.assertEqual(
+                         response.ospfSuperCIDR, "200.100.0.0/18,216.100.0.0/20"
```
Consider extracting the CIDR list to a constant to ensure consistency and increase clarity

jburwell commented on test/integration/smoke/test_ospf_zone_config.py:
```diff
@@ -0,0 +1,165 @@
+        response = self.apiclient.vpcQuaggaConfig(cmd)
+        self.debug("Response quaggaEnabled: %s" % response.quaggaEnabled)
+        self.debug("Response quaggaPassword: %s" % response.quaggaPassword)
+        self.debug("Response ospfSuperCIDR: %s" % response.ospfSuperCIDR)
```
Lines 119-145 and lines 71-95 appear similar.  Consider extracting to a private method.

jburwell commented on utils/src/main/java/com/cloud/utils/net/NetUtils.java:
```diff
@@ -903,6 +906,10 @@ static boolean areCidrsNotEmpty(String cidrA, String cidrB) {
 
+    public static Long[] cidrToLong(final CIDR ocidr) {
+        return cidrToLong(ocidr.toString());
+    }
```
Why is this method implemented here rather than on ``CIDR`` class?

jburwell commented on utils/src/main/java/com/cloud/utils/net/NetUtils.java:
```diff
@@ -1135,6 +1142,21 @@ public static boolean isSameIpRange(final String cidrA, final String cidrB) {
         return false;
     }
 
+    public static boolean validateGuestCidrForOSPF(final CIDR cidr, final CIDR[] ospfSuperCIDRList) {
```
Consider moving the method to the ``Protocol`` enumeration.

jburwell commented on utils/src/main/java/com/cloud/utils/net/NetUtils.java:
```diff
@@ -1159,6 +1181,25 @@ public static boolean validateGuestCidr(final String cidr) {
         }
     }
 
+    public static boolean validateGuestCidr(final CIDR cidr) throws BadCIDRException {
```
Why is this method implemented here rather on ``CIDR`` for IPv4/IPv6 dual stack operation?

abhinandanprateek replied:
it is being used by code other than ospf to check for RFC1918 compliance.

jburwell commented on utils/src/main/java/com/cloud/utils/net/NetUtils.java:
```diff
@@ -1159,6 +1181,25 @@ public static boolean validateGuestCidr(final String cidr) {
+
+        final CIDR cidr1 = CIDRFactory.getCIDR("10.0.0.0/8");
+        final CIDR cidr2 = CIDRFactory.getCIDR("172.16.0.0/12");
+        final CIDR cidr3 = CIDRFactory.getCIDR("192.168.0.0/16");
```
Consider moving these values to a ``final static ImmutableSet`` on the IPv4 CIDR class.  The ``CIDR`` interface could also provide ``getGuestCIDRs`` method.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/BadCIDRException.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.io.IOException;
+
+public class BadCIDRException extends IOException {
```
Consider adding the value of invalid CIDR as member of this class.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/BadCIDRException.java:
```diff
@@ -0,0 +1,33 @@
+
+    private static final long serialVersionUID = -432126076052875403L;
+
+    public BadCIDRException(String msg) {
```
Consider accepting only the bad CIDR value on the constructor and encapsulating the creation of the error message.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+    public boolean equals(Object o);
+
+    public int hashCode();
```
It is unnecessary to declare ``toString``, ``equals``, and ``hashCode`` on an interface since all classes are ancestors of ``Object``.  Therefore, these method declarations can be removed.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+import java.net.InetAddress;
+import com.cloud.utils.Nullable;
+
+public interface CIDR extends Comparable<CIDR>, Nullable {
```
All interface methods on a ``public`` interface are inherently ``public``.  Therefore, the ``public`` modifiers can be removed.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public int getMask() {
+            // TODO Auto-generated method stub
```
Please address and remove the TODO

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public boolean contains(CIDR cidr) {
+            // TODO Auto-generated method stub
```
Please address and remove the TODO

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public boolean overlaps(CIDR cidr) {
+            // TODO Auto-generated method stub
```
Please address and remove the TODO

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public boolean contains(InetAddress cidr) {
+            // TODO Auto-generated method stub
```
Please address and remove the TODO

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public int compareTo(CIDR o) {
+            // TODO Auto-generated method stub
```
Please address and remove the TODO

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+        @Override
+        public InetAddress getBaseAddress() {
+            // TODO Auto-generated method stub
+            return null;
```
Ideally, a Null Object should not return ``null`` instances.  Any thoughts on an ``InetAddress`` instance we could use a logically null reference (e.g. would ``InetAddress`` of 0.0.0.0 make sense?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+
+/**
+ */
+public class CIDR4 implements CIDR {
```
Why not declare this class as ``final``?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    protected final static Logger s_logger = Logger.getLogger(CIDR4.class);
+
+    protected InetAddress baseAddress;
+    protected int cidrMask;
```
Why are these attributes declared ``protected`` rather than ``private``?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    protected InetAddress baseAddress;
+    protected int cidrMask;
+    private int addressInt;
+    private final int addressEndInt;
```
Why aren't all attributes declared ``final``?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+        try {
+            baseAddress = intToIPv4Address(addressInt);
+        } catch (UnknownHostException e) {
+            // this should never happen
```
While this should never happen, it still might.  Throw an ``IllegalStateException`` just in case to fail fast.

abhinandanprateek replied:
Changed it to throw Error.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    public int compareTo(CIDR arg) {
+        if (arg instanceof CIDR6) {
```
Check that ``arg`` is not ``null`` before the ``instanceof`` check to avoid an NPE.  Also, to positively ensure that a ``ClassCastException`` does not occur on line 59, consider changing the ``instanceof`` check to ``!(arg instanceof CIDR4)`` to be explicit about the type.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    @Override
+    public boolean contains(InetAddress inetAddress) {
+        if (inetAddress == null) {
+            throw new NullPointerException("inetAddress");
```
``NullPointerException`` should only be thrown by VM when a null point is dereferenced.  Convert to a ``Preconditions.checkArgument`` check to simplify the code and ensure the correct exception type is thrown.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+            net |= addres & 0xFF;
+        }
+        return net;
+    }
```
Would it make sense to move the ``intToIPv4Address``, ``ipv4AddressToInit``, ``ipv4AddressToInt`` methods to the ``IpAddress`` interface?

Also, consider utilizing covarient types for the ``ipv4AddressToInt`` method -- allowing it to converted to ``int addressToInt(final Inet4Address address)`` and ``int addressToInt(final Inet6Address address)``.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    @Override
+    public boolean contains(CIDR cidr) {
+        // TODO Auto-generated method stub
+        return false;
```
Should method have an implementation?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+        if (!(o instanceof CIDR)) {
+            return false;
+        }
+        return compareTo((CIDR)o) == 0;
```
``equals`` should perform a content based comparison for equality.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof CIDR)) {
```
``instanceof`` is an insufficent check in a multi-classloader environment such as Tomcat.  Please use CID4.class.equals(o.getClass).  Also, check that ``o`` is not ``null`` to avoid an NPE.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    private int addressInt;
+    private final int addressEndInt;
+
+    protected CIDR4(Inet4Address newaddr, int mask) {
```
Default visibility should be sufficient for this constructor.  Also, add a Preconsitions.checkArgument check for ``newAddr`` being ``null``.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+    @Override
+    public boolean overlaps(CIDR cidr) {
+        Long[] cidrALong = NetUtils.cidrToLong(this);
+        Long[] cidrBLong = NetUtils.cidrToLong(cidr);
```
Consider extracting ``NetUtils.cidrToLong()`` to this class and defining ``toLong`` as a method on the ``CIDR`` interface.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR.java:
```diff
@@ -0,0 +1,98 @@
+public interface CIDR extends Comparable<CIDR>, Nullable {
+
+    final static CIDR NULL = new NullCIDR();
+    final static int MAX_CIDR = 32;
```
The ``static`` modifier is unnecessary on a interface declaration because all members are ``static`` by definition.  Therefore, the ``static`` modifiers can be removed.  Also, is the ``MAX_CIDR`` the same for IPv4 and IPv6?

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR4.java:
```diff
@@ -0,0 +1,172 @@
+
+    @Override
+    public int hashCode() {
+        return baseAddress.hashCode();
```
Please include the values of all attributes in the hash code calculation.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDR6.java:
```diff
@@ -0,0 +1,95 @@
+import org.apache.commons.lang.NotImplementedException;
+import org.apache.log4j.Logger;
+
+public class CIDR6 implements CIDR {
```
This class seems to be an empty implementation.  Why are we including it at this time?

abhinandanprateek replied:
It is going to be implemented at some point.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDRFactory.java:
```diff
@@ -0,0 +1,79 @@
+            throw new BadCIDRException("Invalid mask length used: " + maskString);
+        }
+        return getCIDR(addr, mask);
+    }
```
Consider renaming ``getCIDR`` to ``create`` since the ``get`` verb is generally in Java for methods that access attributes.

jburwell commented on utils/src/main/java/com/cloud/utils/net/cidr/CIDRFactory.java:
```diff
@@ -0,0 +1,79 @@
+        } catch (UnknownHostException e) {
+            throw new BadCIDRException("The fomat of address string is not valid " + addr);
+        }
+    }
```
Consider moving this method to the ``IpAddress`` interface as ``createFromHostName``.

jburwell commented:
@abhinandanprateek Reading through the code, it seems like there could be benefit from the following value objects:

* IpAddress interface whose implementations would wrap InetAddress to provide factory methods and operations specific to CloudStack orchestration
* Netmask interface to encapsulate netmask-related operations for IPv4 and IPv6

Assuming that these new network primitives should orient us towards deprecating ``NetUtils``,  the new methods added to ``NetUtils`` for this enhancement should find a place in these new primitives.

abhinandanprateek commented:
@jburwell briefly moving to java 8 made me realise that it has lot more ipv6 primitives then there are in 7.

swill commented:
Please verify why the jobs are failing and re-push this PR to try to get it green.  I need to verify the process to test new system templates and how to build and distribute new system templates.  I have not tried to find the documentation on this, so if someone knows where to find it, I would accept a link.  :)

DaanHoogland commented:
@rhtyd @weizhouapache is this functionality worth keeping an eye on?

weizhouapache commented:
wow, this is a quite big change.
I think it worths re-visiting. @DaanHoogland 
not now, but 4.15 or 4.16.

rhtyd commented:
Hold-on, maybe useful in future when we want to allow route-able networks with bgp/ospf.

rhtyd commented:
Closing this old work, we'll revisit this in future.
